src/utils/constants/autocomplete.ts

```diff
@@ -0,0 +1,94 @@
+import { WasiCommands } from "../../commands/wasiCommands";
+import { SUPPORTED_COMMANDS } from "./commandInfo";
+
+export function autocomplete(commandSnippet: string, commandHistory: string[]): string {
+    const splitArgs = commandSnippet.split(' ');
+    if (commandSnippet.length == 0 || splitArgs[0].length == 0) {
+        //default return help if empty
+        return "help";
+    }
+
+    //is command if all user has written is a command no space
+    const isCommand = splitArgs.length == 1 && !commandSnippet.endsWith(' ');
+    if (isCommand) {
+        return autocompleteCommand(splitArgs[0]);
+    } else {
+        return autocompleteParam(splitArgs, commandHistory);
+    }
+}
+
+function autocompleteCommand(command: string): string {
+    // concat supported commands + compiled commands
+    const dict = Object.keys(SUPPORTED_COMMANDS)
+        .concat(Object.keys(WasiCommands.compiledCommands))
+    
+    const bestRec = getBestRec(command, dict);
+
+    return bestRec.substring(command.length, bestRec.length);
+}
+
+function autocompleteParam(splitArgs: string[], commandHistory: string[]): string {
+    const command = splitArgs[0];
+    const lastArg = splitArgs[splitArgs.length-1]; 
+
+    const paramsList = getAllParamsUsedWithCommand(command, commandHistory);
+    
+    if (paramsList.length == 0) {
+        return "?"; // disregard if unknown -- recognize attempt
+    }
+
+    const bestRec = getBestRec(lastArg, paramsList);
+
+    return bestRec.substring(lastArg.length, bestRec.length);
+}
+
+function getAllParamsUsedWithCommand(command: string, commandHistory: string[]): string[] {
+    let paramList: string[] = [];
```

Comment:
It's cool to have the history search but I think usually terminals map it to a different keystroke than tab. Typically for params tab auto-completes to file names. Maybe as a separate feature, we could list files from the fs provider and auto complete file\folder names - that would certainly help with commands like `code` `cat` etc.

Reply:
Ahhh good point! Not sure why i didn't think of this. I'll update it in this PR and push it out with file autocomplete in mind